File: src/sc_neurocore/accel/julia/neurons/amari_field.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia mirror of the Amari 1977 periodic neural field

"""Source-level Heaviside periodic-grid specialization of Amari equation (3)."""
module AmariFieldAccel

export simulate_amari_field!, validate_amari_field

"""Return whether a complete Amari numerical configuration is in domain."""
function validate_amari_field(
    u::AbstractVector{<:Real}, tau::Real, a_exc::Real, a_width::Real,
    b_inh::Real, b_width::Real, dx::Real, dt::Real,
)::Bool
    return length(u) >= 2 && all(isfinite, u) &&
        all(isfinite, (tau, a_exc, a_width, b_inh, b_width, dx, dt)) &&
        tau > 0 && a_exc >= 0 && a_width > 0 && b_inh >= 0 &&
        b_width > 0 && dx > 0 && dt > 0
end

"""
Advance a caller-owned drive matrix through the Amari field.

Rows of ``currents`` are simultaneous spatial stimuli. ``states_out`` has the
same shape and ``rates_out`` receives active-site fractions. The returned
vector is the final state. Validation and candidate failures throw before a
partial candidate is committed.
"""
function simulate_amari_field!(
    u_init::AbstractVector{<:Real}, tau::Real, a_exc::Real, a_width::Real,
    b_inh::Real, b_width::Real, dx::Real, dt::Real,
    currents::AbstractMatrix{<:Real}, states_out::AbstractMatrix{<:Real},
    rates_out::AbstractVector{<:Real},
)
    validate_amari_field(u_init, tau, a_exc, a_width, b_inh, b_width, dx, dt) ||
        throw(ArgumentError("invalid Amari field configuration"))
    steps, n = size(currents)
    length(u_init) == n || throw(ArgumentError("Amari state/input width mismatch"))
    size(states_out) == (steps, n) || throw(ArgumentError("Amari states_out shape mismatch"))
    length(rates_out) == steps || throw(ArgumentError("Amari rates_out length mismatch"))
    all(isfinite, currents) || throw(ArgumentError("Amari inputs must be finite"))
    u = Float64.(u_init)
    kernel = Vector{Float64}(undef, n)
    for offset in 0:n-1
        distance = min(offset, n - offset) * Float64(dx)
        kernel[offset + 1] = Float64(a_exc) * exp(-Float64(a_width) * distance) -
            Float64(b_inh) * exp(-Float64(b_width) * distance)
    end
    kernel[1] > 0 && kernel[div(n, 2) + 1] < 0 ||
        throw(ArgumentError("Amari kernel is not lateral inhibitory"))
    candidate = similar(u)
    scale = Float64(dt) / Float64(tau)
    @inbounds for step in 1:steps
        for i in 1:n
            convolution = 0.0
            for j in 1:n
                if u[j] > 0.0
                    offset = mod((i - 1) - (j - 1), n)
                    convolution += kernel[offset + 1]
                end
            end
            candidate[i] = u[i] + (-u[i] + convolution * Float64(dx) + Float64(currents[step, i])) * scale
        end
        all(isfinite, candidate) || throw(ArgumentError("invalid Amari candidate state"))
        u .= candidate
        states_out[step, :] .= u
        rates_out[step] = count(>(0.0), u) / n
    end
    return u
end

end # module AmariFieldAccel

File: src/sc_neurocore/accel/julia/neurons/amari_field.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — dedicated Julia neuron facade

from __future__ import annotations

from typing import Any

import numpy as np
import numpy.typing as npt

from ._runtime import JULIA_MAIN as _jl
from ._runtime import KERNEL_DIR as _KERNEL_DIR

_AMARI_FIELD_LOADED = False

def _ensure_amari_field_loaded() -> Any:
    """Include ``amari_field.jl`` into Julia Main on first use."""
    global _AMARI_FIELD_LOADED
    if _jl is None:
        raise ImportError("juliacall not available; install the `julia` extra")
    if not _AMARI_FIELD_LOADED:
        jl_path = _KERNEL_DIR / "amari_field.jl"
        if not jl_path.is_file():
            raise FileNotFoundError(f"amari_field.jl missing at {jl_path}")
        _jl.include(str(jl_path))
        _AMARI_FIELD_LOADED = True
    return _jl.AmariFieldAccel


def simulate_amari_field(
    u_init: npt.ArrayLike,
    tau: float,
    a_exc: float,
    a_width: float,
    b_inh: float,
    b_width: float,
    dx: float,
    dt: float,
    currents: npt.ArrayLike,
) -> dict[str, object]:
    """Run the complete vector batch through the native Julia module."""
    state = np.ascontiguousarray(u_init, dtype=np.float64)
    drive = np.ascontiguousarray(currents, dtype=np.float64)
    if state.ndim != 1 or drive.ndim != 2 or drive.shape[1] != state.size:
        raise ValueError("Julia Amari state/input shape mismatch")
    if not np.isfinite(state).all() or not np.isfinite(drive).all():
        raise ValueError("Julia Amari state and input must be finite")
    states = np.empty_like(drive)
    rates = np.empty(drive.shape[0], dtype=np.float64)
    module = _ensure_amari_field_loaded()
    final = np.asarray(
        module.simulate_amari_field_b(
            state,
            tau,
            a_exc,
            a_width,
            b_inh,
            b_width,
            dx,
            dt,
            drive,
            states,
            rates,
        ),
        dtype=np.float64,
    )
    return {"states": states, "mean_rates": rates, "final_state": final}
